feat(freben): add handling and docs for entity references (#2532)
spotify/backstage: docs/features/software-catalog/references.md

namespace was chosen to fall back to the value `default` by the code that parsed
the reference, so the end result is that we expect to find another entity in the
catalog that is of kind `Group`, namespace `default` (which, actually, also can
be left out in its own yaml file because that's the default value there too),
and name `pet-managers`.

The entries in `implementsApis` are also references. In this case, none of them
need to specify a kind since we know from the context that that's the only kind
that's supported here. The second entry specifies a namespace but the other ones
don't, and in this context, the default is to refer to the same namespace as the
originating entity (`external-systems` here). So the three references
essentially expand to `api:external-systems/petstore`,
`api:internal/streetlights`, and `api:external-systems/hello-world`. We expect
there to exist three API kind entities in the catalog matching those references.

## Compound References

This is a more verbose version of a reference, where each part of the
kind-namespace-name triplet is expressed as a field in a structure. This format

> can be use where necessary, such as if either of the three elements contain

Reviewer: ```suggestion
can be used where necessary, such as if either of the three elements contain
```

> colons or forward slashes. Avoid using its use, since it is harder to read and

Reviewer: ```suggestion
colons or forward slashes. Avoid using it where possible, since it is harder to read and
```

read and write for humans.

```yaml
# Example:
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: petstore
  description: Petstore
spec:
  type: service
  lifecycle: experimental
  owner:
    kind: Group

>     name: aegis-imports/pet-managers

Reviewer 2: Is the actual name of the component here `aegis-imports/pet-managers`? Thought we didn't allow `/` in component names

Author: Just a theoretical example :) Yes this should be rare

Reviewer 2: No I mean we shouldn't accept this as input? It can exist in the file, but we would make `parseEntityRef` or w/e throw?

Author: Not really. The rule of "no slashes" is not fixed, it's configurable. It's perfectly possible that an org makes a custom validator for their entities and allows slashes and/or colons.

Reviewer 2: Seems like a lot of complication for little value to allow that to be configurable

Author: One day we get that customer that goes "oh but our existing inventory and external systems has slashes all over the place, what do you mean you chose to hard code not supporting that?" :P it kinda grinds me to even consider deciding on a thing like that, usually it turns out to not be done for particularly good reason :)

Author: (being a bit stoic here, i know, but that's how it's built already actually)

Reviewer 2: I don't think the plugin ecosystem is gonna support that though, not unless we're way more clear that that is the case

Author: I think it's just up to basic sound engineering practices right? Like encodeURIComponent when building urls etc

Reviewer 2: Yeah but that's not done throughout the codebase atm
